src/stream/src/executor/backfill/cdc/cdc_backfill.rs

```diff
@@ -434,13 +437,29 @@ impl<S: StateStore> CdcBackfillExecutor<S> {
                     }
                 }
             }
+        } else if self.disable_backfill {
```

Comment:
I have a general question: will the states in the cdc backfill state table be ever read if backfill is disabled? If not, is it better to create a `ChainExecutor` instead of a `CdcBackfillExecutor` that skip all the backfill logic?

Reply:
https://github.com/risingwavelabs/risingwave/pull/14718#issuecomment-1909436180

Agree, that's what I'm thinking as well.

Reply:
I think you mean a `ChainExecutor` with `upstream_only=true`. As I mentioned [here](https://github.com/risingwavelabs/risingwave/pull/14718#discussion_r1469646004), we also need to transform upstream data which in `payload jsonb` schema to table columns. The upstream of `CdcBackfillExecutor` is a `CdcFilter` executor not the `Source` by design.

Reply:
I see. I thought there is an explicit project executor after filter executor and before backfill executor. I think I am confused by the naming of `CdcBackfillExecutor` and thought it was only doing backfill. In fact, this exeuctor takes up the following responsibilities:
1. Perform snapshot read on upstream source table
2. Transform jsonb payload data from upstream binlog into payload data with correct table schema
3. Filter out event from upstream binlog with offset < last bin log offset to resolve the race condition between starting snapshot read and starting changelog read.

With `snapshot = 'false'`, 1 can be skipped and 3 can be ignore because the race doesn't exist if 1 is skipped. However, since 2 is still required, we need `CdcBackfillExecutor` instead of a simple `ChainExecutor`.